folly/hash/Checksum.cpp

```diff
@@ -86,7 +96,48 @@ bool crc32_hw_supported() {
   return id.sse42();
 }
 
-#else
+#elif FOLLY_ARM_FEATURE_CRC32
+uint32_t crc32_hw(const uint8_t* buf, size_t len, uint32_t crc) {
+  auto* buf_64 = reinterpret_cast<const uint64_t*>(buf);
```

Comment:
This looks like unaligned access.

Note that while an architecture may support unaligned access, the language generally deems unaligned access of this form to be undefined behavior. The kind of undefined behavior that may end up being subject to miscompilations.

To avoid this unaligned access, do we need to start off with an optional round of each of `__crc32b`, `__crc32h`, `__crc32w`, and `__crc32d`? Alternatively, perhaps we can `memcpy` from `buf` instead of `reinterpret_cast` - modern compilers recognize this idiom and lower to `mov` or `ldr` instructions without emitting calls to `memcpy`.

Here and `crc32c_hw` below.

Reply:
Thanks, you're totally right. Unaligned access isn't directly forbidden in C++ but my implementation is incorrect by the strict aliasing rule. While this is *typically fine* when operating on bytes my code is still incorrect. 

The memcpy version is a good idea. It's obviously correct and at a first glance seems to lead to (nearly) the same code for GCC/clang. I'll test it and update the PR.